UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -845,6 +850,22 @@ ParseDtb (
   for (Node = FdtNextNode (Fdt, 0, &Depth); Node >= 0; Node = FdtNextNode (Fdt, Node, &Depth)) {
     NodePtr = (FDT_NODE_HEADER *)((CONST CHAR8 *)Fdt + Node + Fdt32ToCpu (((FDT_HEADER *)Fdt)->OffsetDtStruct));
     DEBUG ((DEBUG_INFO, "\n   Node(%08x)  %a   Depth %x", Node, NodePtr->Name, Depth));
+
+    // Ensure we don't submit a child of reserved-memory as main memory block.
+    NodeType = CheckNodeType (NodePtr->Name, Depth);

```

Comment:
I think this should be better way to handle duplicate memory nodes inside and outside reserved-memory. With this the old workaround can be removed - CheckMemoryNodeIfInit() and RecordMemoryNode()

Reply:
I was working on a feature recently where I did something that might be better (if I understand the issue correctly): we initialised a cursor in the main loop, and passed it in as an in-out pointer to each function. They'd do their work, then adjust the cursor to the end of the block they handled. That way, we maintained a strictly forward moving, linear walk over the data.

In this case, we wouldn't have to worry about a nested `memory@` block inside `reserved-memory`, `ParseReservedMemory()` would handle it, then adjust our node number.